File: BOJ/5212.cpp
#include <bits/stdc++.h>
using namespace std;

// 좌표를 저장하기 위한 class
class Point {
public:
	int x;
	int y;

	Point(int x, int y) {
		this->x = x;
		this->y = y;
	}
};

int R, C;
Point sp(10, 10); // 50년 후의 지도를 그리기 위한 직사각형의 왼쪽 위 좌표
Point ep(0, 0); // 50년 후의 지도를 그리기 위한 직사각형의 오른쪽 아래 좌표
int dx[4] = { 0, 0, -1, 1 };
int dy[4] = { -1, 1, 0, 0 };
vector<string> Map;

void Solve() {
	vector<string> tmp = Map;
	for (int y = 0; y < R; y++) {
		for (int x = 0; x < C; x++) {
			if (Map[y][x] == '.') continue; // 현재 위치가 바다인 경우 신경쓸 필요가 없다.

			int seaCount = 0;
			for (int i = 0; i < 4; i++) {
				int nx = x + dx[i];
				int ny = y + dy[i];

				// 현재 노드 기준으로 상,하,좌,우 4방향을 살펴보고 바다(.)이면 카운트를 +1 해준다.
				if (nx < 0 || nx >= C || ny < 0 || ny >= R || Map[ny][nx] == '.') seaCount++;
			}
			if (seaCount >= 3) tmp[y][x] = '.'; // 3방향 이상이 바다이면 50년 후 섬은 잠긴다.
			else { // 섬이 남아 있다면 50년 후의 지도를 그리기 위해 최대 좌표(직사각형) 값을 업데이트 해준다.
				sp.x = min(sp.x, x);
				sp.y = min(sp.y, y);
				ep.x = max(ep.x, x);
				ep.y = max(ep.y, y);
			}
		}
	}

	// output
	for (int i = sp.y; i <= ep.y; i++) {
		for (int j = sp.x; j <= ep.x; j++) {
			cout << tmp[i][j];
		}
		cout << '\n';
	}
}

void Input() {
	string s;
	cin >> R >> C;
	for (int i = 0; i < R; i++) {
		cin >> s;
		Map.push_back(s);
	}
}

int main() {

	Input();
	Solve();

	return 0;
}

File: BOJ/python/5212.py
import sys

input = sys.stdin.readline

R, C = map(int, input().split(' '))
Map = [str(input().rstrip('\n')) for i in range(R)]
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]
sx = sy = 10 # 50년 후의 지도를 기리기 위한 직사각형의 왼쪽 위 좌표
ex = ey = 0 # 50년 후의 지도를 기리기 위한 직사각형의 오른쪽 아래 좌표

tmp = [[1 if Map[y][x] == 'X' else 0 for x in range(C)] for y in range(R)]
for y in range(R):
    for x in range(C):
        if Map[y][x] == '.': # 현재 위치가 바다일 경우 넘어간다.
            continue

        sea_count = 0

        # 현재 위치 기준 상, 하, 좌, 우 4방향을 살펴보고 바다이면 카운트를 +1 해준다.
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if (nx < 0) or (nx >= C) or (ny < 0) or (ny >= R) or (Map[ny][nx] == '.'):
                sea_count += 1

        # 3방향이 바다이면 50년 후 섬은 잠긴다.
        if sea_count >= 3:
            tmp[y][x] = 0

        # 섬이 남아 있다면 50년 후의 지도를 그리기 위해 최대 사각형의 좌표 값을 업데이트 해준다.
        else:
            sx = min(sx, x)
            sy = min(sy, y)
            ex = max(ex, x)
            ey = max(ey, y)

# 50년 후의 지도 출력
for y in range(sy, ey + 1):
    print(''.join(['X' if tmp[y][x] == 1 else '.' for x in range(sx, ex + 1)]))
